cuda/src/memory/cuda/device_memory_resource.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#include "../../utils/cuda_error_handling.hpp"
+
+#include "vecmem/memory/resources/memory_resource.hpp"
+#include "vecmem/memory/cuda/device_memory_resource.hpp"
+
+#include <cuda_runtime_api.h>
+
+namespace vecmem::cuda {
+    void * device_memory_resource::do_allocate(
+        std::size_t bytes,
+        std::size_t
+    ) {
+        void * res;
+        VECMEM_CUDA_ERROR_CHECK(cudaMalloc(&res, bytes));
```

Comment:
With `m_device` added, you should add

```c++
VECMEM_CUDA_ERROR_CHECK(cudaSetDevice(m_device));
```

before the `cudaMalloc(...)` call.

Reply:
Agreed. However, we also have the option of returning the CUDA device to what is was before the allocation, so that the call to this function serves as a sort of stack, and the CUDA device is reset when it exits. What do you think is best? Personally I like the idea of setting it back to what it was at the end of the call, since it reduces the number of outwardly observable side-effects.

Reply:
Sure, I didn't even think of that. But I guess in the end the safest would be something like:

```c++
static std::mutex s_memAllocMutex;
...
void * device_memory_resource::do_allocate( std::size_t bytes, std::size_t ) {
   std::lock< std::mutex > lock( s_memAllocMutex );
   int currentDevice = 0;
   VECMEM_CUDA_ERROR_CHECK( cudaGetDevice( &currentDevice ) );
   VECMEM_CUDA_ERROR_CHECK( cudaSetDevice( m_device ) );
   void* res = nullptr;
   VECMEM_CUDA_ERROR_CUECK( cudaMalloc( &res, bytes ) );
   VECMEM_CUDA_ERROR_CUECK( cudaSetDevice( currentDevice ) );
   return res;
}
```

I.e. we would even need to use a mutex to avoid parallel threads doing something unexpected.

Reply:
In fact, the CUDA device selector API is already thread-safe and thread-local, so this should not be necessary!